### PR TITLE
Remove the longhorn-upgrade attacher type and refactor the upgrade path

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#6139

### CODE REVIEW

**Comment by PhanLe1010** [comment]
Waiting for e2e test run before marking this PR ready for review

**Review by innobead — APPROVED** [review]
LGTM

File: upgrade/v14xto150/upgrade.go [hunk 1]
```diff
@@ -116,25 +121,46 @@ func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset, resourceM
...
-		return errors.Wrapf(err, "failed to list all existing Longhorn volumes during the VolumeAttachment upgrade")
+		return errors.Wrapf(err, "failed to list all existing Longhorn volumes during the Longhorn VolumeAttachment upgrade")
+	}
+	kubeVAList, err := kubeClient.StorageV1().VolumeAttachments().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all Kubernetes VolumeAttachments during the Longhorn VolumeAttachment upgrade")
+	}
+	smList, err := lhClient.LonghornV1beta2().ShareManagers(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all pods in %v during the Longhorn VolumeAttachment upgrade", namespace)
```
Comment (innobead) [inline upgrade/v14xto150/upgrade.go:139]:
```suggestion
		return errors.Wrapf(err, "failed to list all share managers in %v during the Longhorn VolumeAttachment upgrade", namespace)
```

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Review by derekbit — APPROVED** [review]
LGTM

**Review by shuo-wu — APPROVED** [review]

File: upgrade/v14xto150/upgrade.go [hunk 2]
```diff
@@ -21,6 +22,10 @@ const (
 )
 
 func UpgradeResources(namespace string, lhClient *lhclientset.Clientset, kubeClient *clientset.Clientset, resourceMaps map[string]interface{}) error {
+	if err := upgradeCSIPlugin(namespace, kubeClient); err != nil {
```
Comment (innobead) [inline upgrade/v14xto150/upgrade.go:25]: Why we need to do this specially? isn't it handled by driver deployment instead?

Comment (shuo-wu) [reply upgrade/v14xto150/upgrade.go:25]: You can check [item 2 in this comment](https://github.com/longhorn/longhorn/issues/6139#issuecomment-1596457985) for details

**Comment by PhanLe1010** [comment]
@innobead The test is still running here https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4256/

**Comment by PhanLe1010** [comment]
The test result looks good. There some failing one but they doesn't seems related to the PR. Except for the one `test_migration_with_unscheduled_replica`. I am retesting `test_migration_with_unscheduled_replica` manually

**Comment by PhanLe1010** [comment]
Update: `test_migration_with_unscheduled_replica` passed when testing manually. Maybe the test case was a bit flaky. 

Suggesting merging the PR. Thank you very much @innobead

**Comment by innobead** [comment]
cc @longhorn/qa for the flaky test case

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Pull request merged**